src/epics.c

```diff
@@ -0,0 +1,551 @@
+    if (p->value.values == NULL) {
+      ERROR("epics plugin: Cannot allocate memory for %lu values or channel "
+            "\"%s\"",
+            ds->ds_num, p->name);
+
+      return;
+    }
+  }
+
+  ret =
+      ca_create_subscription(p->ch_type, ca_element_count(p->id), p->id,
+                             DBE_VALUE | DBE_ALARM, event_handler, p, &p->eid);
+  if (ret != ECA_NORMAL) {
+    ERROR("epics plugin: CA error %s occurred while trying to create "
+          "subscription for channel \"%s\"",
+          ca_message(ret), p->name);
+
+    return;
+  }
+
+  p->is_active = 1;
+}
+
+static void handle_conn_down(struct pv *p) {
+  WARNING("epics plugin: Channel \"%s\" disconnected", p->name);
+
+  p->is_active = 0;
+}
+
+static void connection_handler(struct connection_handler_args args) {
+  struct pv *p = (struct pv *)ca_puser(args.chid);
+
+  switch (args.op) {
+  case CA_OP_CONN_UP:
+    handle_conn_up(p);
+    break;
+  case CA_OP_CONN_DOWN:
+    handle_conn_down(p);
+    break;
+  }
+}
+
+static void *epics_thread(void *args) {
+  int i = 0;
+  long ret = 0;
+
+  ret = ca_context_create(ca_disable_preemptive_callback);
+  if (ret != ECA_NORMAL) {
+    // FIXME: report error back to start_thread()
+    ERROR("epics plugin: CA error %s occurred while trying to start channel "
+          "access",
+          ca_message(ret));
+    return (void *)1;
+  }
+
+  ca_replace_printf_handler(&printf_handler);
+
+  for (i = 0; i < epics_plugin.pvs_num; ++i) {
+    struct pv *p = &epics_plugin.pvs[i];
+
+    ret = ca_create_channel(p->name, &connection_handler, p, 0, &p->id);
+    if (ret != ECA_NORMAL) {
+      ERROR("epics plugin: CA error %s occurred while trying to create channel "
+            "\"%s\"",
+            ca_message(ret), p->name);
+      ret = 1;
+      goto error;
+    }
+  }
+
+  while (epics_plugin.thread_loop != 0) {
+    ca_pend_event(2.0);
+  }
+
+error:
+  for (; i > 0; --i) {
+    struct pv *p = &epics_plugin.pvs[i - 1];
+
+    if (p->eid) {
+      ca_clear_subscription(p->eid);
+    }
+    ca_clear_channel(p->id);
+  }
+
+  ca_context_destroy();
+
+  return (void *)ret;
+}
+
+static int start_thread(void) {
+  int ret = 0;
+
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  if (epics_plugin.thread_loop != 0) {
+    goto epics_unlock;
+  }
+
+  epics_plugin.thread_loop = 1;
+  ret = plugin_thread_create(&epics_plugin.thread_id, epics_thread, (void *)0,
+                             "epics");
+  if (ret != 0) {
+    epics_plugin.thread_loop = 0;
+    ERROR("epics plugin: Starting thread failed: %d", ret);
+
+    goto epics_unlock;
+  }
+
+  // FIXME: wait untill ca_context_create success
+
+epics_unlock:
+  pthread_mutex_unlock(&epics_plugin.lock);
+
+  return ret;
+}
+
+static int stop_thread(void) {
+  int ret = 0;
+
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  if (epics_plugin.thread_loop == 0) {
+    goto epics_unlock;
+  }
+
+  epics_plugin.thread_loop = 0;
```

Comment:
You can skip the conditional and the goto. Just set the variable unconditionally, which has the same effect.

```suggestion
  epics_plugin.thread_loop = 0;
```